### PR TITLE
LPD-24485 Fix JSONWebServiceActionsManager concurrent issue exposed by JDK21

### PR DESCRIPTION
**Pull request by julschong**
@tinatian
For issue: https://liferay.atlassian.net/browse/LPD-24485

I have testing PR here isolating the failed KBUpgrade functional test: https://github.com/julschong/liferay-portal/pull/1226

### CODE REVIEW

**Comment by julschong** [comment]
ci:test:relevant

**Comment by julschong** [comment]
@tinatian no regression

File: modules/apps/portal-remote/portal-remote-json-web-service-web/src/main/java/com/liferay/portal/remote/json/web/service/web/internal/JSONWebServiceActionsManagerImpl.java [hunk 1]
```diff
@@ -313,7 +313,7 @@ private int _countMatchedParameters(
 		return matched;
 	}
 
-	private void _ensureOpen() {
+	private synchronized void _ensureOpen() {
```
Comment (dantewang) [inline modules/apps/portal-remote/portal-remote-json-web-service-web/src/main/java/com/liferay/portal/remote/json/web/service/web/internal/JSONWebServiceActionsManagerImpl.java:316]: @julschong If one thread triggers `_serviceTracker.open()` here, then all other threads calling any of this class' public API should be blocked by the `synchronized(this)` in the `getSingleton()` method. Why can't the `synchronized` in `getSingleton` protect from this race condition?

Comment (julschong) [reply modules/apps/portal-remote/portal-remote-json-web-service-web/src/main/java/com/liferay/portal/remote/json/web/service/web/internal/JSONWebServiceActionsManagerImpl.java:316]: @dantewang As far as why adding `synchronized` to the method works, I'm not 100% certain. And need your help on this.
My concern is that the lock on the method should be released when the thread exits the method allowing 1 other thread to continue. but in reality, first thread never releases the lock until it is done populating the map

**Comment by tinatian** [comment]
@julschong Close as we talked, please resend.

Tina

**Comment by julschong** [comment]
resubmitted https://github.com/liferay-core-infra/liferay-portal/pull/6928

File: modules/apps/portal-remote/portal-remote-json-web-service-web/src/main/java/com/liferay/portal/remote/json/web/service/web/internal/JSONWebServiceActionsManagerImpl.java (hunk 1 above)
Comment (julschong) [reply modules/apps/portal-remote/portal-remote-json-web-service-web/src/main/java/com/liferay/portal/remote/json/web/service/web/internal/JSONWebServiceActionsManagerImpl.java:316]: @dantewang I looked at it again and debugged further. The issue is that serviceTracker.open() is run twice.
see resubmit: https://github.com/liferay-core-infra/liferay-portal/pull/6928

**Pull request closed without merging**